### PR TITLE
Refactor parseTreeEntries, speed up tree list

### LINKED ISSUE #20315: [BUG]:  slice bounds out of range at modules/git
### Description

I tried to parse some data with ParseTreeEntries in modules/git, but it crashed instead of returning an error

```go
package main
import "code.gitea.io/gitea/modules/git"
func main() {
	git.ParseTreeEntries([]byte("1006440000000000000000000000000000000000000000000000000000000000000000000000\t\n1006440000000000000000000000000000000000000000000000000000000000000000000000\t\n1006440000000000000000000000000000000000000000000000000000000000000000000000\t\n1006440000000000000000000000000000000000000000000000"))
}

```
Found by [go-fuzz](https://github.com/dvyukov/go-fuzz)

### Gitea Version

with git commit as ee769f7459ffba310563cd735f503d3d7576da60

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

Ubuntu 

### How are you running Gitea?

I test the API in Gitea through go-fuzz

### Database

_No response_

### PR DESCRIPTION
**Pull request by wxiaoguang**
This PR will:
* Close #20315
* Speed up #20231

```
$ time git ls-tree -t -r HEAD | wc -l
81941
git ls-tree -t -r HEAD  0.11s user 0.08s system 75% cpu 0.252 total
wc -l  0.01s user 0.00s system 5% cpu 0.251 total

$ time git ls-tree -t -r -l HEAD | wc -l
81941
git ls-tree -t -r -l HEAD  1.00s user 0.77s system 38% cpu 4.578 total
wc -l  0.03s user 0.00s system 0% cpu 4.579 total
```

### CODE REVIEW

File: services/repository/files/tree.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ func GetTreeBySHA(ctx context.Context, repo *repo_model.Repository, gitRepo *git
 	tree.URL = repo.APIURL() + "/git/trees/" + url.PathEscape(tree.SHA)
 	var entries git.Entries
 	if recursive {
-		entries, err = gitTree.ListEntriesRecursive()
+		entries, err = gitTree.ListEntriesRecursive("--long") // get blob sizes
```
Comment (delvh) [inline services/repository/files/tree.go]: Alternatively, we could use two methods, one with and one without the size calculation.

**Review by lunny — APPROVED** [review]

**Review by delvh — APPROVED** [review]

File: services/repository/files/tree.go (hunk 1 above)
Comment (wxiaoguang) [reply services/repository/files/tree.go]: Good suggestion. At the moment only 2 calls to it. Maybe next time we can think a better new name (I admit that I can not think a new name for it now 😁 )

Comment (wxiaoguang) [reply services/repository/files/tree.go]: Maybe
* ListEntriesRecursiveFast
* ListEntriesRecursiveWithSize
?

**Comment by wxiaoguang** [comment]
bot

**Pull request merged**